### PR TITLE
Add `sub_ptr` on pointers (the `usize` version of `offset_from`)

### PR DESCRIPTION
**Pull request by scottmcm**
We have `add`/`sub` which are the `usize` versions of `offset`, this adds the `usize` equivalent of `offset_from`.  Like how `.add(d)` replaced a whole bunch of `.offset(d as isize)`, you can see from the changes here that it's fairly common that code actually knows the order between the pointers and *wants* a `usize`, not an `isize`.

As a bonus, this can do `sub nuw`+`udiv exact`, rather than `sub`+`sdiv exact`, which can be optimized slightly better because it doesn't have to worry about negatives.  That's why the slice iterators weren't using `offset_from`, though I haven't updated that code in this PR because slices are so perf-critical that I'll do it as its own change.

This is an intrinsic, like `offset_from`, so that it can eventually be allowed in CTFE.  It also allows checking the extra safety condition -- see the test confirming that CTFE catches it if you pass the pointers in the wrong order.

### CODE REVIEW

**Comment by Gankra** [comment]
Sorry for the drive-by comment (rant?) but just wanted to leave the note somewhere: since i've been poking these APIs, this has been the second-most-brain-melty interface in `ptr` (align_offset of being the undefeated champion).

* offset_from refuses to say the math expression it's actually expressing (is it `x-y` or `y-x`??)
* you don't actually use these APIs when you care about `offset`, you care about them when you want to know the distance
* `_from` is totally backwards and makes things feel double or triple negative

Like my guess is that the intent is that `x.offset_from(y)` is a drop-in for any existing `x - y` but: the docs don't say that and instead appeal to `offset`, and any time I was doing `x - y` I was *already* struggling to hold the right way around in my brain, and most appealing to "well you want unsigned semantics, so you don't want underflow, so put the big one first". offset_from scuppers this because it's explicitly signed... so yeah having unsigned APIs would be super great for helping with that bit!

This API is (at least in std) pretty universally used for computing array length via pointer distance... and yes you do this with `end - start` but that's an implementation detail and not actually what you're trying to think about. You really are trying to ask "how many items are in the range start..end", and subtract is the means to that end. Reproducing the "double negative" of putting things backwards is *understandable* but I think missing an opportunity.

imo things end up a lot clearer if you just make this `start.offset_to(end)` because it's like, yeah that's what I'm always doing. I offset from the start to the end, which is always positive. And I offset to, not from.

And then yeah you always want it unsigned so... what you really want is `unsigned_offset_to`. But this is an Objectively Better API In Every Way to `offset_from` (just as add is objectively better than offset), so it should be shorter (just as add is shorter than offset), and, oh, hey right we have add.

So, just make `add_to`, IMO. 😸

(I will concede add_to reads funny, because my brain wants to map it to "add x to y" and not "the add to get to y", but I think this is a weird enough name that it will just... be fine in practice and we'll get over it in like a day when all of our code actually makes sense to read again.)

Also CC #95643

**Comment by Gankra** [comment]
(also yeah byte_unaligned_offset_from is Suffering, byte_add_to is uh, a lot more reasonable lol)

**Comment by RalfJung** [comment]
> _from is totally backwards and makes things feel double or triple negative

FWIW I find the naming very intuitive -- `x.offset_from(y)` is the offset you need to add to `y` to reach `x`, i.e., it is `x`'s "distance when starting from" `y`. I didn't struggle with operand order when using it. Maybe by brain is wired in weird ways :shrug: .

I do agree that it should explicitly spell out the `(y-x) / size` in the docs, but that seems like a minor thing.

I agree `offset_to` would also work. `add_to` on the other hand sounds very confusing to me, `x.add_to(y)` should be `y+x` IMO.

File: compiler/rustc_const_eval/src/interpret/intrinsics.rs [hunk 1]
```diff
@@ -348,16 +348,30 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
                         let usize_layout = self.layout_of(self.tcx.types.usize)?;
                         let isize_layout = self.layout_of(self.tcx.types.isize)?;
+                        let ret_layout = if intrinsic_name == sym::ptr_offset_from {
+                            isize_layout
+                        } else {
+                            usize_layout
+                        };
                         let a_offset = ImmTy::from_uint(a_offset.bytes(), usize_layout);
                         let b_offset = ImmTy::from_uint(b_offset.bytes(), usize_layout);
                         let (val, _overflowed, _ty) =
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/intrinsics.rs:370]: Hm, I feel like the existing code here made little sense? We want the subtraction to happen in `isize` with the signed version, right?

So this should probably all use the same layout, the one you now call `ret_layout`. (Not that it makes any difference for the result, thanks to two's complement.)

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/intrinsics.rs:370]: In fact, looking at the safety docs, we should probably use `isize` for `a_offset` and `b_offset`, and then actively check for overflow here?

**Comment by eddyb** [comment]
(I know @Gankra already knows this, but I wanted to state my objection with "add" because it's only a verb, whereas AFAICT `offset_from` uses "offset" as a noun, not as a verb like the `offset` method does. So `offset_from` is more like `{quantity}_from` than `{action}_from`)

Personally something like `start.distance_to(end)`, or `(start..end).element_len()` seems potentially straight-forward enough to be acceptable (with `start` and `end` both being pointers - and `element_` could be replaced by `byte_` for the byte-oriented one).
<sub>(I do not have the same issue tying `offset_from` to subtraction, but I've failed to figure out why it seems "natural" to think of "directional distances" in this way, and if it doesn't for other people I don't mind replacing it with anything that's more range-like, which have much "wider" precedent in Rust than raw pointers)</sub>

**Comment by saethlin** [comment]
Just caught up on the Zulip thread for this.

@scottmcm you commented:

> Huh, I should have rg'd sooner -- essentially every use of ptr::offset_of is followed by as usize:

I think `add` and `sub` were a huge improvement over `offset`, because they are both clearer and less characters to type. They're an unambiguous win over code that does `offset(usize as isize)`. Data suggests that people only really want the `usize` version of `ptr - ptr`, so I think it would be a shame if we had an API which is not clearer but has a longer name to do the operation people want.

**Comment by est31** [comment]
Yeah `offset` absolutely suffers from the verb / noun confusion issue, where the two mean completely different things. To give a data point, when I see `v.offset_from(w)`, I think about the result being something like `w + v`, not about it being `v - w`.

`distance` suffers from the verb/noun issue less, even though usages as a verb exist. As the API is stable since 1.47 it's tough to change it, but of course expansion of the API can provide a good opportunity to revisit the choice and maybe provide a new one with a better name, deprecating the old one, and the added function only using the new name.

Regarding `offset_to` vs `offset_from`, I like the fact that the latter is analog to the `-` operator. I feel that Rust should not make the abstractions even less natural than they have to be due to the provenance issue. I think most people work with a mental model of pointers as numbers, and Rust should respect that.

**Comment by Gankra** [comment]
I opened https://github.com/rust-lang/rust/pull/95851 to clarify offset_from's docs

**Comment by RalfJung** [comment]
> Regarding offset_to vs offset_from, I like the fact that the latter is analog to the - operator. I feel that Rust should not make the abstractions even less natural than they have to be due to the provenance issue. I think most people work with a mental model of pointers as numbers, and Rust should respect that.

If we want an operation that works in `const`, we *have to* require that both pointers point to the same allocation. Allocations at `const` time do not have fixed addresses, so computing offsets across allocation boundaries is not meaningful.

This is also slightly helpful for compiler optimizations, for roughly the same reason -- the optimizer can be sure that passing the result of an `offset_from` to outside code will not leak any information about the allocation base address, which helps analyses. However, arguably, with `x.addr() - y.addr()`, we now have another way to express this that avoids the "same allocation" pitfall. So if we ignore `const`, having a fully safe version is possible without optimization impact. (That's assuming that we keep `addr` with the semantics from https://github.com/rust-lang/rust/pull/95588, and it is theoretic until the compiler starts treating `addr` and `expose_addr` differently.)

**Comment by eternaleye** [comment]
One thing I'll note is that both `end.offset_from(start)` and `start.offset_to(end)` require that you care about the ordering in advance, but many use cases _don't_ care about order - just the (magnitude of) the distance, not the sign.

In such cases, something like `x.distance(y)` such that `start.distance(end) = end.distance(start)` is considerably more ergonomic, and the sign can always be recovered _afterwards_ using `bool negative = x < y`.

**Comment by tschuett** [comment]
You can either recover the sign at runtime  or  have different return types.

**Comment by scottmcm** [comment]
> i.e., it is x's "distance when starting from" y.

I like the `sub` direction, because it's what you'd get doing this with indexes.  I guess we could lean on this in the name too, like `b.sub_ptr(a) = n  <=>  a.add(n) = b  <=>  b.sub(n) = a`.  (That uses a suffix to describe the argument in the same sort of way as [`checked_add_unsigned`](https://doc.rust-lang.org/nightly/std/primitive.i16.html#method.checked_add_unsigned).)

> Data suggests that people only really want the usize version of ptr - ptr, so I think it would be a shame if we had an API which is not clearer but has a longer name to do the operation people want.

:+1:  The same as how `.add(d)` is shorter than `.offset(i)` because it's what I always wanted anyway, it makes sense to me to say that this should have a shorter name than `offset_from` -- and definitely not a much longer one like `unsigned_offset_from`.

> also yeah byte_unsigned_offset_from is Suffering

Heh, very much so.

Hmm, `offset_from` is anticommutative (`a.offset_from(b) == -b.offset_from(a)`), which probably makes the order fundamentially hard to express well.

Maybe we can take advantage of the "UB in the other order (unless they're equal)" to put the directionality more obviously in the name.

Would it be clearer with `b.bytes_after(a)` and `b.elements_after(a)`?  Though including `elements` seems somewhat inconsistent with the `add` and `sub` methods that aren't `add_bytes` and `add_elements`...

> but many use cases _don't_ care about order - just the (magnitude of) the distance, not the sign.

Can you say more about this, @eternaleye?  Notably, every single case in the `library/` seems to know up-front which is larger than the other, and @saethlin's case [above](https://github.com/rust-lang/rust/pull/95837#issuecomment-1094069283) does too.  So while an [`abs_diff`](https://doc.rust-lang.org/std/primitive.i32.html#method.abs_diff)-like API certainly seems reasonable, it's not obvious to me that it should be the primary one.

**Comment by Gankra** [comment]
I think @eternaleye was arguing that the order-less API is *strictly more useful* because it works for both the standard "definitely know" and the strange "no idea (yet)". But I think this operation is low-level enough and the "definitely know" case so common, that we should prefer the ordered impl because it's more optimized (saving like 3 instructions or something, I couldn't think of any way to do cute hardware-specific).

Also we have [usize::abs_diff](https://doc.rust-lang.org/std/primitive.usize.html#method.abs_diff) so if anyone *really* wants it they can do `ptr.addr().abs_diff(ptr2.addr()) / size` themselves and it will probably be negligible since they're doing something Weird.

**Comment by scottmcm** [comment]
Thanks, @Gankra; that makes sense.  And we definitely need a hyper-optimized version of this operation for use [in the bowels of slice iterators](https://github.com/rust-lang/rust/blob/8c1fb2eb23964b56bd279b12aa925935d8559870/library/core/src/slice/iter/macros.rs#L25-L35), so it might as well be this method.  I agree the more unusual cases are best left for something else.

---

I ended up pushing a change to rename to `sub_ptr`, as the consistency with the `add`/`sub` names resonated the strongest with me.  [This test](https://github.com/rust-lang/rust/pull/95837/files#diff-94ddab85eeb9ffede9cb938e799d366a4e13acaa09a13a0cbdd8992ab1a0d9d6R50) in particular emphasized it the strongest -- if I look at `ptr.add(2).sub_ptr(ptr)` it seems perfectly clear to me for that to return `2`, just like it would for `i + 2 - i`.

I also added a section like this to the docs to attempt to further emphasize:
```rust
ptr.sub_ptr(origin) == count
origin.add(count) == ptr
ptr.sub(count) == origin
```

Feedback appreciated on whether this name is good, or whether you're prefer one of the other naming approaches.

**Comment by Gankra** [comment]
I still think framing it as subtraction is a missed opportunity because subtraction is not what you actually "care" about, but a means to an end (which always stops me in my tracks while I triple check which way I want).

But at least it is very compact and appeals directly to subtraction, so it's a strict ergonomics win with the same intuition as the "normal" way. I would happily let this land to nightly and let the name bikeshed later. 😸

**Comment by Gankra** [comment]
(disclaimer: I have not reviewed the rest of this in detail, and idk how much the docs want to crib from my other PR)

File: compiler/rustc_const_eval/src/interpret/intrinsics.rs (hunk 1 above)
Comment (scottmcm) [reply compiler/rustc_const_eval/src/interpret/intrinsics.rs:370]: That's true.  Probably both versions should always do the subtraction in `isize`, so the overflow flag will check the `isize::MAX` limit.

I can make that change, but is there a way I can write a test for it?  I would guess that CTFE won't actually let me make an allocation big enough to overflow the limit without failing the "different allocations" check first.

**Comment by saethlin** [comment]
Could we have a debug assertion for getting the arguments in the wrong order? I think unlike `offset_from` there is a wrong order for a function that returns `usize`.

---
You could use `core::intrinsics::assert_unsafe_precondition` if you need. I would prefer we not need, but the impact on code size and compile time is undeniable and it automates the dance around not doing anything in CTFE because not everything is a `const fn`.

**Comment by scottmcm** [comment]
> You could use core::intrinsics::assert_unsafe_precondition if you need.

Ah, I'll try that.  I'd actually included a `debug_assert!` in one of the things I pushed today (<https://github.com/rust-lang/rust/runs/5957231198?check_suite_focus=true#step:26:221>), but removed it after getting the "pointers cannot be reliably compared during const eval" error.

**Comment by RalfJung** [comment]
> (Although, thinking about this again, a.offset_from(b).unsigned_abs() works fine since the the signed distance isn't allowed to wrap anyway, and thus it's less obvious that a separate method for it is needed, the way abs_diff is.)

(FWIW I think we might [want to allow lather distances in the future](https://github.com/rust-lang/rust/issues/92512#issuecomment-1094081032).)

**Comment by saethlin** [comment]
What is a lather distance?

**Comment by scottmcm** [comment]
I'm guessing "larger" distance (`rg` are right next to `th`).

Technically, the LLVM implementation for `ptr_offset_from_unsigned` in this PR would work perfectly fine without the "no more than `isize::MAX` apart" precondition, since it's working entirely in unsigned operations and thus there are no overflow complications.

**Comment by RalfJung** [comment]
As long as we require both pointers to be inbounds of the same allocation, they cannot be more than `isize::MAX` apart anyway as that is the maximal size of an allocation.

File: compiler/rustc_const_eval/src/interpret/intrinsics.rs [hunk 2]
```diff
@@ -348,16 +348,37 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+                        let (val, overflowed, _ty) =
                             self.overflowing_binary_op(BinOp::Sub, &a_offset, &b_offset)?;
+                        if overflowed {
+                            throw_ub_format!("Pointers were too far apart for {}", intrinsic_name);
+                        }
+
                         let pointee_layout = self.layout_of(substs.type_at(0))?;
-                        let val = ImmTy::from_scalar(val, isize_layout);
-                        let size = ImmTy::from_int(pointee_layout.size.bytes(), isize_layout);
+                        let val = ImmTy::from_scalar(val, ret_layout);
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/intrinsics.rs:380]:
```suggestion
                        // This re-interprets an isize at ret_layout, but we already checked
                        // that if ret_layout is usize, then the result must be non-negative.
                        let val = ImmTy::from_scalar(val, ret_layout);
```

File: library/core/src/ptr/const_ptr.rs [hunk 3]
```diff
@@ -611,6 +611,83 @@ impl<T: ?Sized> *const T {
...
+    ///     assert_eq!(ptr1.add(2), ptr2);
+    ///     assert_eq!(ptr2.sub(2), ptr1);
+    ///     assert_eq!(ptr2.sub_ptr(ptr2), 0);
+    /// }
+    ///
+    /// // This would be incorrect, as the pointers are not correctly ordered:
+    /// // ptr1.offset_from(ptr2)
+    /// ```
+    #[unstable(feature = "ptr_unsigned_offset_from", issue = "88888888")]
+    #[rustc_const_unstable(feature = "const_ptr_offset_from", issue = "92980")]
```
Comment (RalfJung) [inline library/core/src/ptr/const_ptr.rs:675]: `const_ptr_offset_from` is in FCP and will be stabilized soon. I think constness of this new function should be tracked together with its non-const stability.

**Comment by scottmcm** [comment]
I've added a real tracking issue (#95892) and resolved all the open review comments.

@rustbot ready

**Comment by RalfJung** [comment]
Why did you use two different feature gates for the `fn` and `const fn`?
I don't think that is necessary.

**Comment by scottmcm** [comment]
> I don't think that is necessary.

I have no idea whether it is, so I just used two because I know that works and most of the const ones are `const_*`.

**Comment by RalfJung** [comment]
Yeah but that's usually because we are constifying things that are already stable, or making this const fundamentally extends what const code can do. Neither is the case here.

**Comment by scottmcm** [comment]
Hello libs-api folks!  This has an LGTM for the code, but is waiting on your approval for the (unstable) name.

This API can replace `offset_from` in every single use I found, so I think it should be a relatively-terse one, like how we now almost always use `.add(i)` instead of `.offset(i as isize)`.

My reasoning for using the name I did is in https://github.com/rust-lang/rust/pull/95837#issuecomment-1094132593, but I'm happy to change it to whatever.

**Comment by scottmcm** [comment]
Rebase from faee850 to e1520f1 is just updating to master.
Rebase from e1520f1 to 003b954 is [fixing the codegen test](https://github.com/rust-lang/rust/compare/e1520f17142fdd3fe0482ee5a4d984f25fb0262c..003b954a43a7f1f9058f25e8f9b6ddfd4a3dced9).

There's an `alloca` in the bors run that doesn't show up locally for me, but it looks like -O1 will fix that: https://rust.godbolt.org/z/rjoxYWErc

@bors r=oli-obk rollup=iffy (codegen tests are always iffy for me)

**Pull request merged**